### ISSUE TITLE
Rate this session feature

### ISSUE THREAD

**Issue by eleanorreem** [chaynHQ/bloom-frontend#858]
### Overview
We want feedback for our content so we can do better. We are implementing a "rate this session" feature on all our sessions.

**Full page layout:**
<img width="393" alt="Screenshot 2024-02-20 at 12 23 29" src="https://github.com/chaynHQ/bloom-frontend/assets/16049515/5682211a-0194-4697-be10-d2e0b49abcd2">

**Desktop designs:** 
<img width="368" alt="Screenshot 2024-02-20 at 12 21 44" src="https://github.com/chaynHQ/bloom-frontend/assets/16049515/f4c98bc5-88e2-47ca-836a-724fd8688c9a">

<img width="550" alt="Screenshot 2024-02-20 at 12 57 29" src="https://github.com/chaynHQ/bloom-frontend/assets/16049515/751be1f3-f902-4ca5-8938-b875a1f6e079">

**Mobile designs:**
<img width="266" alt="Screenshot 2024-02-20 at 12 21 52" src="https://github.com/chaynHQ/bloom-frontend/assets/16049515/4d7c9826-8e2d-48fe-b140-fca94fcbeccd">

<img width="286" alt="Screenshot 2024-02-20 at 12 29 49" src="https://github.com/chaynHQ/bloom-frontend/assets/16049515/fa4a16d9-3aa2-44d9-ad56-5513da92449b">

Please note this ticket is blocked by this [backend ticket](https://github.com/chaynHQ/bloom-backend/issues/396)

### Action Items
- [ ] If you would like access to the figma files for more exact spacing - please ping @eleanorreem. However, it doesn't have to be pixel perfect but getting it as close to the eye as possible. If you have any responsiveness questions, please comment below with any questions and tag @eleanorreem
- [ ] Create the new form in /components/forms
- [ ] Add the form to the page /pages/courses/[slug]/[sessionSlug].tsx
- [ ] Add the text strings to the /messages/courses folder. Fill all the non english files with english and this will be translated at a later date before we release the feature. 
- [ ] When the user submits - you need to hit the new POST /api/v1/session-feedback
- [ ] Show the success screen. If there are any failures, for now, we will just show the success screen. Ensure the error is logged in rollbar and then we can follow up with a fix. 
- [ ] Update the test /cypress/integration/user-course-session-behaviour.cy.tsx. You should add a step to fill out the feedback form and ensure that the test results are correct. 

### Resources/Instructions
- As much as possible, please use components already made. If you need to add variations on a component, please do. 
- Our form components are currently stored in /components/forms. Please store this component there too.
- Please reuse the checkbox and form field components used on the /components/forms/AboutYouDemographicForm.tsx. This form can be used as a guide on how to structure your forms.

**Comment by ckirby19** [comment]
Hi, once my PR for the blocking backend ticket is merged, I would like to start on this one, thank you :)

**Comment by eleanorreem** [comment]
Brill thanks @ckirby19 ⭐

**Comment by ckirby19** [comment]
Hi, I have made progress on this ticket (see below image of current look) but had the following issue I am stuck on:

I need the sessionId to send to the backend and my current attempt at doing this can be found here: https://github.com/chaynHQ/bloom-frontend/commit/809aeec032111337af7c723edd8446f1d2418c0a#diff-1f886e305706be9b6675ef7f50d14cd4e2253dc2f81291ab9e6f9075f95d366aR69-R92

However, I find that const courses = useTypedSelector((state) => state.courses) is empty and so I have no courses to use to try and get the sessionId. Is there another way of getting this Id that I am missing? Thanks!

![SessionFeedback](https://github.com/user-attachments/assets/d1a94175-39c8-4d2b-a90d-e7f898a90406)

**Comment by eleanorreem** [comment]
Hey @ckirby19 great progress. I had a look into the code and you are right that the page is not pulling the course/ session data.

I had a look and the course data is only being loaded in these scenarios below:
<img width="696" alt="Screenshot 2024-07-17 at 15 59 24" src="https://github.com/user-attachments/assets/d84f527e-e9f1-4466-b1a7-a35bd7d984df">

i.e.  when a user is created/ get user details on page load/ completes a session/  starts a session. So the courses state actually represents the courseUser relation rather than a full list of courses and sessions. You only get an entry in the courseUser table if you actually interact with the session by clicking play on a video or looking at the transcript. I hope that makes sense.

Have you clicked play on any videos or opened any videos on that session? Because if you have then my theory could be wrong!

I looked at the other instances where we update an entity that requires the sessionId. They all rely on the fact that the course state has been populated i.e. the user has already interacted with the Session. Or it relies on the storyblokId. 

See [SessionProgressDisplay](https://github.com/chaynHQ/bloom-frontend/blob/develop/components/session/SessionProgressDisplay.tsx) for an instance of relying on storyblokId and state. 

See [SessionCompleteButton](https://github.com/chaynHQ/bloom-frontend/blob/develop/components/session/SessionCompleteButton.tsx) where it relies on the backend to handle storyblokId. 

I see a few options. Sadly some require some work on the backend:

1. The POST /session-feedback endpoint has to be able to handle storyblokId and know about the sessionEntity. 
2. We fetch the session data from the backend by hitting a GET /session endpoint on page load which you can pass a searchCriteria query like GET /user. However this endpoint doesn't exist... 😬
3. You could serve an error requiring the user to start the session by clicking the video or transcript before submitting their feedback. Once they have watched the session, the session will exist on course state and you should be able to submit it. Not sure whether I like this option as its making the user work to fix your problem. Although you would expect them to start a session before they review 🤷  

There are probably more solutions to this problem but there is some food for thought. I'll leave the choice to you!

**Comment by eleanorreem** [comment]
Just had some feedback from the product manager. She would be totally fine with the scenario where we require users to start the session before allowing them to give feedback. So if you haven't already gone down another route, we could throw up an error on the form saying "Please watch a video or read a transcript before providing feedback". That can be a placeholder now and we can improve the copy later.

**Comment by ckirby19** [comment]
Hi @eleanorreem, I actually just saw this after raising a PR haha. So the approach I took (and let me know if this seems like a problem), was to move the userSession checking logic to the [sessionSlug.tsx] file, and then to selectively render the feedback form only if the userSession is defined (which as you correctly stated above, will only occur if the user has started or finished the course). If it is preferred that we instead always show the feedback page but give an error to the user if they try to submit it without starting a course, I am happy to do that instead. Thanks!
